src/scenic/core/sample_checking.py

```diff
@@ -0,0 +1,108 @@
+"""The SampleChecker class and it's implementations."""
+
+from abc import ABC, abstractmethod
+import time
+
+from scenic.core.requirements import IntersectionRequirement
+
+class SampleChecker(ABC):
+    def __init__(self):
+        self.requirements = None
+
+    def addRequirements(self, requirements):
+        assert self.requirements is None
+        self.requirements = tuple(requirements)
+
+    @abstractmethod
+    def checkRequirementsInner(self, sample):
+        pass
+
+    def checkRequirements(self, sample):
+        assert self.requirements is not None
+        return self.checkRequirementsInner(sample)
+
+class BasicChecker(SampleChecker):
+    """ Basic requirement checker.
+    
+    Evaluates requirements in order, with a tiny bit of tuning.
+    """
+    def __init__(self, initialCollisionCheck):
+        super().__init__()
+        self.initialCollisionCheck = initialCollisionCheck
+
+    def addRequirements(self, requirements):
+        target_reqs = []
+        for req in requirements:
+            if req.optional:
+                # Basic checker ignores optional requirements unless otherwise noted.
+                if (self.initialCollisionCheck and
+                    sum(isinstance(r, IntersectionRequirement) for r in requirements) >= 3):
+                    target_reqs.append(req)
+            else:
+                target_reqs.append(req)
+
+        super().addRequirements(target_reqs)
+
+    def checkRequirementsInner(self, sample):
+        for req in self.requirements:
+            if req.active and req.falsifiedBy(sample):
+                return req.violationMsg
+
+        return None
+
+class WeightedAcceptanceChecker(SampleChecker):
+    """ Picks the requirement with the lowest time-weighted acceptance chance.
+
+    Incentivizes exploration by initializing all buffer values to 0.
+
+    Args:
+        bufferSize: Max samples to use when calculating time-weighted
+            rejection chance.
+    """
+    def __init__(self, bufferSize=10):
+        super().__init__()
+        self.bufferSize = bufferSize
+
+    def addRequirements(self, requirements):
+        super().addRequirements(requirements)
+
+        self.buffers = {req:[(1,0)]*self.bufferSize for req in self.requirements}
+
+    def checkRequirementsInner(self, sample):
+        for req in self.sortRequirements():
+            # Evaluate the requirement with timing info.
+            start = time.time()
```

Comment:
Should use `time.perf_counter()` for these times since it's monotonic (and in theory `time()` can have very poor resolution). I'm guilty of using `time()` elsewhere -- we should probably go through and fix that!

Reply:
Ah very cool, I didn't know this existed!